src/wp-includes/class-wp-script-modules.php

```diff
@@ -359,4 +366,27 @@ private function get_src( string $id ): string {
 
 		return $src;
 	}
+
+	public function print_script_data(): void {
+		$modules = array();
+		foreach ( array_keys( $this->get_marked_for_enqueue() ) as $id ) {
+			$modules[ $id ] = true;
+		}
+		foreach ( array_keys( $this->get_import_map()['imports'] ) as $id ) {
+			$modules[ $id ] = true;
+		}
+
+		foreach ( array_keys( $modules ) as $module_id ) {
+			$config = apply_filters( 'scriptmoduledata_' . $module_id, array() );
+			if ( ! empty( $config ) ) {
+				wp_print_inline_script_tag(
+					wp_json_encode( $config, JSON_HEX_TAG | JSON_HEX_AMP ),
```

Comment:
Any reason not to include [`JSON_UNESCAPED_SLASHES`](https://www.php.net/manual/en/json.constants.php#constant.json-unescaped-slashes) here too? I suspect URLs of endpoints will be relatively common data, and avoiding the [leaning toothpick syndrome](https://en.wikipedia.org/wiki/Leaning_toothpick_syndrome) would be nice. Having `JSON_HEX_TAG` should be sufficient to avoid problems with `</script>` that the default escaping of slashes was added to avoid.

[`JSON_UNESCAPED_UNICODE`](https://www.php.net/manual/en/json.constants.php#constant.json-unescaped-unicode) is another one to consider, but that would need testing for whether it breaks if pages are served with an encoding other than UTF-8.

Reply:
I didn't consider this at all, it's copied from how the Interactivity API implements the same thing:

https://github.com/WordPress/wordpress-develop/blob/473e2554db8e547a07a16f73080ca49c0c30b89f/src/wp-includes/interactivity-api/class-wp-interactivity-api.php#L170-L179

I did some testing and it seems like we should be able to add `JSON_UNESCAPED_SLASHES` in both of these places, thanks!

Reply:
`JSON_UNESCAPED_UNICODE` might also be nice to add, but as you mention it might need to be behind a UTF-8 check:

```
$opts = JSON_HEX_TAG | JSON_HEX_AMP | JSON_UNESCAPED_SLASHES;
if ( get_bloginfo( 'blog_charset' ) === 'UTF-8' ) {
  $opts |= JSON_UNESCAPED_UNICODE;
}
```

It looks like it would certainly improve readability and potentially size:

```php
<?php
$opts = 0;
$t = '🏴󠁧󠁢󠁥󠁮󠁧󠁿';
var_dump(
    $t,
    strlen($t),
    json_encode($t, $opts),
    strlen(json_encode($t, $opts)),
    json_encode($t, $opts | JSON_UNESCAPED_UNICODE),
    strlen(json_encode($t, $opts | JSON_UNESCAPED_UNICODE)),
);
```

```
string(28) "🏴󠁧󠁢󠁥󠁮󠁧󠁿"
int(28)
string(86) ""\ud83c\udff4\udb40\udc67\udb40\udc62\udb40\udc65\udb40\udc6e\udb40\udc67\udb40\udc7f""
int(86)
string(30) ""🏴󠁧󠁢󠁥󠁮󠁧󠁿""
int(30)
```

Reply:
I'll follow up for Interactivity API here: https://github.com/WordPress/wordpress-develop/pull/6520

Reply:
In some quick experimenting, I find Firefox 125 and Chromium 124 both do need the document to be utf-8 to interpret unescaped Unicode properly.

Also they seem to use the actual `Content-Type` HTTP header over `<meta>` tags if both specify a different charset, but they'll use `<meta>` tags if the HTTP header doesn't specify a charset. The deprecated `charset` attribute on a `<script>` tag doesn't seem to be used by either.

Reply:
It is vital that `JSON_UNESCAPED_SLASHES` not be added, as it avoids a security vulnerability when encoding a string that includes `</script>`. When slashes are escaped, this becomes `<\/script>` and the browser won't prematurely close the script tag.

Reply:
Thanks for flagging that! I'd like to continue discussing on https://github.com/WordPress/wordpress-develop/pull/6520#issuecomment-2102224570, where I've replied. That PR is likely to land sooner, and whatever implementation is used there should also be used here. (In fact, I'd expect Interactivity API to use this proposal to print its data.)

Reply:
@westonruter I already addressed that concern

> Having `JSON_HEX_TAG` should be sufficient to avoid problems with `</script>` that the default escaping of slashes was added to avoid.

Is there any reason to think that's not accurate?

Reply:
Ah, yes. As long as `JSON_HEX_TAG` is present, then it seems safe to include `JSON_UNESCAPED_SLASHES`. Maybe add a comment explaining why the particular set of flags are used, noting the security implications.